realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.kt

```diff
@@ -128,12 +128,13 @@ class ClassMetaData(env: ProcessingEnvironment, typeMirrors: TypeMirrors, privat
             return type != "io.realm.DynamicRealmObject" && !type.endsWith(".RealmObject") && !type.endsWith("RealmProxy")
         }
 
+    var embedded: Boolean = false
```

Comment:
This is probably an architectural problem. Right now `ClassMetaData` is responsible for both parsing AND storing state. It would probably help if we could separate this a bit. The issue is that there is a lot of state and it is interdependent, so not sure how much it would help if we split it. We would still need to track a lot of intermediate state somewhere.

Reply:
It feels like it could escalate quickly if these changes are done here and now. I don't think it's super urgent so we can push it for a later time. I'm very wary when it comes to using mutable properties in Kotlin nonetheless. If not managed correctly they become a timebomb.

Reply:
If you look at the rest of this class there are tons of var as well. That doesn't mean your point is invalid, but I wonder how you would handle the case of gathering ~10+ variables across 100's of line of code before you send them to a constructor?

Out of

Reply:
I can see e.g. `internalClassName` is a latent var. It takes its value from `realmClassAnnotation` or `moduleClassNameFormatter` (it's a bit more cumbersome to reason in case the value comes from the latter, but still, it's the same rationale as it could be passed via constructor/a builder from where the class metadata is created in `RealmProcessor`). In turn, the local `realmClassAnnotation` variable is immutable and takes its value from `classType` which comes from the constructor as immutable too. All these things considered, there is no excuse to make `internalClassName` a mutable property.

It is true that you might end up having a gazillion properties in your constructor, to which I would strongly object too. You could use a builder to handle that, but that adds a ton of boilerplate. I think the best solution is to encapsulate all these immutable properties behind a data structure/class. This way you would at least keep the clutter away from `ClassMetaData` and those properties will be easily accessible via `val`s with backing properties from `ClassMetaData`, e.g.

```
class ClassMetaData(private val myMetadaContainer: MetaDataContainer) {
   private val internalClassName: String
      get() = myMetadaContainer.internalClassName
```

I don't think there's a silver bullet here in any case but makes the solution more robust IMO.

Reply:
I have no doubt there are things we can do to improve this class. You mentioned some of them, but I would rather not do it in this PR since it is too big already. I created https://github.com/realm/realm-java/issues/6891 to track it.

Reply:
I agree, seems too much for the scope of this PR.